File: core/crates/primitives/src/testkit/transfer_data_extra_mock.rs
use super::signer_mock::TEST_EVM_RECIPIENT;
use crate::swap::ApprovalData;
use crate::{TransactionType, TransferDataExtra, TransferDataOutputAction, TransferDataOutputType};

impl TransferDataExtra {
    pub fn mock() -> Self {
        TransferDataExtra {
            to: TEST_EVM_RECIPIENT.to_string(),
            gas_limit: None,
            gas_price: None,
            data: None,
            output_type: TransferDataOutputType::EncodedTransaction,
            output_action: TransferDataOutputAction::Sign,
            transaction_type: TransactionType::SmartContractCall,
            approval: None,
        }
    }

    pub fn mock_encoded_transaction(data: Vec<u8>) -> Self {
        TransferDataExtra { data: Some(data), ..Self::mock() }
    }

    pub fn mock_signature(data: Vec<u8>, approval: Option<ApprovalData>) -> Self {
        TransferDataExtra {
            data: Some(data),
            output_type: TransferDataOutputType::Signature,
            transaction_type: TransactionType::Transfer,
            approval,
            ..Self::mock()
        }
    }
}

File: android/gemcore/src/testFixtures/kotlin/com/gemwallet/android/testkit/TransferDataExtraMock.kt
package com.gemwallet.android.testkit

import com.gemwallet.android.ext.toGem
import com.wallet.core.primitives.TransactionType
import com.wallet.core.primitives.TransferDataOutputAction
import com.wallet.core.primitives.TransferDataOutputType
import uniffi.gemstone.ApprovalData
import uniffi.gemstone.TransferDataExtra
import java.math.BigInteger

fun mockTransferDataExtra(
    to: String = "recipient",
    data: ByteArray? = null,
    outputType: TransferDataOutputType = TransferDataOutputType.EncodedTransaction,
    outputAction: TransferDataOutputAction = TransferDataOutputAction.Send,
    transactionType: TransactionType = TransactionType.Transfer,
    gasLimit: BigInteger? = null,
    approval: ApprovalData? = null,
) = TransferDataExtra(
    to = to,
    gasLimit = gasLimit,
    gasPrice = null,
    data = data,
    outputType = outputType.toGem(),
    outputAction = outputAction.toGem(),
    transactionType = transactionType.toGem(),
    approval = approval,
)
